src/coreclr/src/System.Private.CoreLib/src/System/ActivatorCache.CoreCLR.cs

```diff
@@ -0,0 +1,126 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.CompilerServices;
+
+namespace System
+{
+    internal sealed partial class RuntimeType
+    {
+        /// <summary>
+        /// A cache which allows optimizing <see cref="Activator.CreateInstance"/>,
+        /// <see cref="RuntimeType.CreateInstanceDefaultCtor"/>, and related APIs.
+        /// </summary>
+        private sealed unsafe class ActivatorCache
+        {
+            // The managed calli to the newobj allocator, plus its first argument (MethodTable*).
+            // In the case of the COM allocator, first arg is ComClassFactory*, not MethodTable*.
+            private readonly delegate*<void*, object?> _pfnAllocator;
+            private readonly void* _allocatorFirstArg;
+
+            // The managed calli to the parameterless ctor, taking "this" (as object) as its first argument.
+            private readonly delegate*<object?, void> _pfnCtor;
+            private readonly bool _ctorIsPublic;
+
+#if DEBUG
+            private readonly RuntimeType _originalRuntimeType;
+#endif
+
+            internal ActivatorCache(
+                [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors | DynamicallyAccessedMemberTypes.NonPublicConstructors)] RuntimeType rt)
+            {
+                Debug.Assert(rt != null);
+
+#if DEBUG
+                _originalRuntimeType = rt;
+#endif
+
+                // The check below is redundant since these same checks are performed at the
+                // unmanaged layer, but this call will throw slightly different exceptions
+                // than the unmanaged layer, and callers might be dependent on this.
+
+                rt.CreateInstanceCheckThis();
+
+                RuntimeTypeHandle.GetActivationInfo(rt, forGetUninitializedInstance: false,
+                    out _pfnAllocator!, out _allocatorFirstArg,
+                    out _pfnCtor!, out _ctorIsPublic);
+
+                Debug.Assert(_pfnAllocator != null);
+                Debug.Assert(_allocatorFirstArg != null);
+
+                if (_pfnCtor == null)
+                {
+                    if (!RuntimeTypeHandle.IsValueType(rt))
+                    {
+                        // Reference type with no parameterless ctor.
+                        // Is it __ComObject? If not, we can't continue.
+
+#if FEATURE_COMINTEROP
```

Comment:
It still does not feel right to have this logic here. I had to step through this to figure out how this condition is compensating for the COM-interop logic on the unmanaged side.

Here is my suggestion for what having all logic on the unmanaged side would look like: https://github.com/jkotas/runtime/commit/953f99c1c3cad958e81d8f3ba41bfc5dafe6d602 (feel free to cherry pick this or edit it as necessary if you like it)

Reply:
The reason I didn't want to introduce that logic on the unmanaged side is that I'm going to have to refactor it yet again once https://github.com/dotnet/runtime/issues/36194 comes online. For instance - the lack of a parameterless might no longer be seen as an error.

As it stands now, I think I'm going to have to reintroduce the _fGetUninitializedObject_ parameter as part of that work item, as using _ppfnCtor == nullptr_ will no longer be a reliable check. Adding the throw logic to the unmanaged method now means I'll need to delete or move it later.

Reply:
Yes, I think it would be best

Reply:
It might make sense to move the "return a null allocator for `Nullable<T>`" behavior back down into the unmanaged code, just like how we have the unmanaged code responsible for finding the COM activation stub. But I'm still concerned about making that single unmanaged method responsible for all the different use cases (activator w/ default constructor, activator w/o default constructor, get uninitialized object).

What if we split the unmanaged method again into two: one that's solely responsible for finding the allocator, and one that's solely responsible for locating the default ctor? Then on the `ActivatorCache` side it'd look something like:

```cs
/* STEP 1: Find the allocator */

RuntimeTypeHandle.GetAllocator(rt, forGetUninitializedInstance: false, out _pfnAllocator, out _vAllogatorArg);

if (_pfnAllocator == null)
{
    Debug.Assert(rt.IsNullable, "Null allocator should only be returned for Nullable<T>.");
    _pfnAllocator = &NullStub;
}

/* STEP 2: Find the default ctor */

RuntimeTypeHandle.GetDefaultCtor(rt, out _pfnDefaultCtor, out _fDefaultCtorIsPublic);

if (_pfnDefaultCtor == null)
{
    // We don't call the ctor for value types and __ComObject, only call the allocator
    if (!rt.IsValueType && !rt.IsComBaseType)
    {
        throw new MethodMissingException(/* ... */);
    }

    _pfnDefaultCtor = &NoopStub;
    _fDefaultCtorIsPublic = true;
}
```

This way, once the other `ConstructorInfo.CreateDelegate` methods come online, they can reuse the logic from `RuntimeTypeHandle.GetAllocator` _with no changes needed to the unmanaged side_.

An earlier prototype had the unmanaged methods split like this, but it was also doing things like passing the _MethodTable\*_ around. These APIs don't transmit the method table pointer, so the split should be cleaner and should leak fewer implementation details.

Thoughts?

Reply:
> What if we split the unmanaged method again into two

It gets non-obvious which one of the two methods is responsible for error handling.

> once the other ConstructorInfo.CreateDelegate methods come online

I expect that this one will have to have quite a bit of logic on the unmanaged side to validate the signature match, etc. And it won't need to deal with any of the odd cases like Nullable because Nullable or fake COM objects without constructors. So the only part that it shares with the Activator path are:
- Error handling: This can be moved into a separate method on the unmanaged side.
- Method to get the allocator. This method is factored out already on the unmanaged side.

I am fine with moving things around again once this method comes online and it can be seen that it fits together nicely. We are talking about ~50 lines moved around, so it should not be a big deal.

Reply:
Resolved offline - we'll split the unmanaged methods into 3, with error handling refactored out into a standalone shared method.